core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -66,140 +72,221 @@ class LogManager(val logDirs: Array[File],
+      def value = offlineLogDirs.length
+    }
+  )
+
   /**
    * Create and check validity of the given directories, specifically:
    * <ol>
    * <li> Ensure that there are no duplicates in the directory list
    * <li> Create each directory if it doesn't exist
-   * <li> Check that each path is a readable directory 
+   * <li> Check that each path is a readable directory
    * </ol>
    */
-  private def createAndValidateLogDirs(dirs: Seq[File]) {
+  private def createAndValidateLogDirs(dirs: Seq[File], initialOfflineDirs: Seq[File]): ArrayBuffer[File] = {
     if(dirs.map(_.getCanonicalPath).toSet.size < dirs.size)
-      throw new KafkaException("Duplicate log directory found: " + logDirs.mkString(", "))
-    for(dir <- dirs) {
-      if(!dir.exists) {
-        info("Log directory '" + dir.getAbsolutePath + "' not found, creating it.")
-        val created = dir.mkdirs()
-        if(!created)
-          throw new KafkaException("Failed to create data directory " + dir.getAbsolutePath)
+      throw new KafkaException("Duplicate log directory found: " + dirs.mkString(", "))
+
+    val liveLogDirs = ArrayBuffer.empty[File]
+
+    for (dir <- dirs if !initialOfflineDirs.contains(dir)) {
+      try {
+        if (!dir.exists) {
+          info("Log directory '" + dir.getAbsolutePath + "' not found, creating it.")
+          val created = dir.mkdirs()
+          if (!created)
+            throw new KafkaException("Failed to create data directory " + dir.getAbsolutePath)
+        }
+        if (!dir.isDirectory || !dir.canRead)
+          throw new KafkaException(dir.getAbsolutePath + " is not a readable log directory.")
+        liveLogDirs += dir
+      } catch {
+        case t: Throwable => error(s"Failed to create or validate data directory $dir.getAbsolutePath", t)
       }
-      if(!dir.isDirectory || !dir.canRead)
-        throw new KafkaException(dir.getAbsolutePath + " is not a readable log directory.")
     }
+    if (liveLogDirs.isEmpty)
+      throw new KafkaException("None of the specified log dirs from " + dirs.mkString(", ") + " can be created or validated")
+
+    liveLogDirs
   }
-  
+
+  def handleLogDirFailure(dir: String) {
+    if (!logDirs.exists(_.getAbsolutePath == dir))
+      throw new RuntimeException(s"Log dir $dir is not found in the config.")
+
+    if (!liveLogDirs.exists(_.getAbsolutePath == dir)) {
+      info(s"Log dir $dir is already offline.")
+      return
+    }
+
+    logCreationOrDeletionLock synchronized {
+      if (!liveLogDirs.exists(_.getAbsolutePath == dir)) {
+        info(s"Log dir $dir is already offline.")
+        return
+      }
+      info(s"Stopping serving logs in dir $dir")
+
+      liveLogDirs -= new File(dir)
+      if (liveLogDirs.isEmpty) {
+        fatal(s"Shutdown broker because all log dirs in ${logDirs.mkString(", ")} have failed")
+        Runtime.getRuntime().halt(1)
+      }
+
+      val offlineTopicPartitions = logs.filter(_._2.dir.getParent == dir).map(_._1)
```

Comment:
Sure. I have replaced this line with `val offlineTopicPartitions = logs.filter { case (tp, log) => log.dir.getParent == dir}.map { case (tp, log) => tp}`